core/src/main/java/com/linecorp/armeria/internal/common/GracefulConnectionShutdownHandler.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.EventLoop;
+
+/**
+ * Abstract class that's used to implement protocol-specific graceful connection shutdown logic.
+ *
+ * <p>This class is <b>not</b> thread-safe and all methods should be called from a single thread such
+ * as {@link EventLoop}.
+ */
+public abstract class GracefulConnectionShutdownHandler {
+    private static final Logger logger = LoggerFactory.getLogger(GracefulConnectionShutdownHandler.class);
+
+    @Nullable
+    private ChannelPromise promise;
+    // Drain duration in microseconds used during the graceful connection shutdown start.
+    private long drainDurationMicros;
+    private boolean canCallOnDrainStart = true;
+    @Nullable
+    private ScheduledFuture<?> drainFuture;
+
+    protected GracefulConnectionShutdownHandler(long drainDurationMicros) {
+        setDrainDurationMicros(drainDurationMicros);
+    }
+
+    private void setDrainDurationMicros(long drainDurationMicros) {
+        checkArgument(drainDurationMicros >= 0, "count: %s (expected: >= 0", drainDurationMicros);
```

Comment:
Agree! Initially I thought that it's good to have this validation to make sure future changes don't break the invariant. But since we don't have any special handling for negative values in this class anymore - I think it's fine to remove the validation, negative value here won't break the internal logic.